File: skills/plotting/python/pavlab_boxplot.py
"""pavlab_boxplot — opinionated lab-style boxplot.

Used **sparingly**. The box itself — quartiles + whiskers + median —
IS the spread/uncertainty visualization; no separate overlaid error
bar is drawn. Two modes:

  - ``show_points=True`` (default): raw data points (swarmed) under
    the box. Points show the actual distribution; the box gives a
    structural summary.
  - ``show_points=False``: compact box only (quartiles + whiskers +
    median). Used in multi-panel figures where visual compactness
    matters and the box's own whiskers carry the spread information.

There is intentionally no ``error=`` parameter: a separate
mean ± SD / ± 95% CI bar overlaid on top of a box would double up the
spread cue the box already encodes via its whiskers and IQR, and the
lab convention is that the box itself does that job. Helpers for
parameter-estimate plots (forest plots, regression-coefficient charts,
etc.) WILL carry an error-bar API in the future, following the
cross-plot rule documented in ``feedback_error_bars_never_sem.md`` —
SD for raw-data charts, 95% CI for parameter-estimate charts, never
SEM.

Usage
-----
    from pavlab_boxplot import pavlab_boxplot

    pavlab_boxplot(groups, values, ylabel="Expression")
    pavlab_boxplot(groups, values, show_points=False)        # compact
    pavlab_boxplot(groups, values, points_kind="strip")      # jittered
"""

from __future__ import annotations

import warnings
from typing import Any, Union

import matplotlib.pyplot as plt
import numpy as np

try:
    from .pavlab_scatter import (
        _pick_log_kind,
        _apply_log,
        _axis_label,
        _font_sizes,
        _maybe_warn_log,
        _resolve_color,
        _N_ALPHA,
    )
    from .pavlab_stripchart import _swarm_x_offsets
except ImportError:
    from pavlab_scatter import (  # script / sys.path style
        _pick_log_kind,
        _apply_log,
        _axis_label,
        _font_sizes,
        _maybe_warn_log,
        _resolve_color,
        _N_ALPHA,
    )
    from pavlab_stripchart import _swarm_x_offsets

# Style constants — match the rest of the plotting skill / CLAUDE.md.
_BOX_COLOR    = "#1f2937"   # gray-800 (TEXT) for box outline + median
_POINT_COLOR  = "#374151"   # gray-700 — slightly lighter than the box so the
                            # box reads as the structural element on top of
                            # the raw points (which are the "data layer").
_BOX_FILL     = "#f3f4f6"   # gray-100 — very pale box fill so the points
                            # underneath stay readable.
_BOX_LW       = 1.2
_MEDIAN_LW    = 1.6
_WHISKER_LW   = 1.0
_CAP_LW       = 1.0
_BOX_WIDTH    = 0.55

LogScale = Union[bool, str]


# ---------------------------------------------------------------------------
# Main function
# ---------------------------------------------------------------------------

def pavlab_boxplot(
    x,
    y,
    color=None,
    color_label: str | None = None,
    show_points: bool = True,
    points_kind: str = "swarm",
    jitter: float = 0.2,
    box_width: float = _BOX_WIDTH,
    order: list[str] | None = None,
    label_size: str | float = "med",
    xlabel: str | None = None,
    ylabel: str | None = None,
    title: str | None = None,
    log_y: LogScale = False,
    pseudocount: float = 1.0,
    origin_zero: bool = True,
    ylim: tuple | None = None,
    point_size: float | None = None,
    filename=None,
    figsize: tuple | None = None,
    ax=None,
) -> Any:
    """Pavlab-style boxplot with raw-points underlayer and error-bar overlay.

    Parameters
    ----------
    x : array-like of str
        Group labels, one per observation.
    y : array-like of float
        Continuous values, one per observation.
    color : None | str | array-like, optional
        Point colour: None → gray-700; single color string → uniform;
        array of strings → categorical colors + legend; array of floats
        → viridis colorbar. Only the *point* layer respects ``color`` —
        the box itself always uses the lab structural colour.
    color_label : str, optional
        Colorbar label for continuous color.
    show_points : bool
        Draw raw data points underneath the box. Default True — the
        lab default. Set False for a compact box-only chart; the
        whiskers and IQR still convey the spread.
    points_kind : "strip" | "swarm"
        Layout for the points underlayer. ``"swarm"`` (default) packs
        points non-overlappingly via the same beeswarm algorithm as
        ``pavlab_stripchart(kind="swarm")``. ``"strip"`` uses uniform
        horizontal jitter.
    jitter : float
        Half-width of horizontal jitter when ``points_kind="strip"``.
        Default 0.2. Ignored under swarm.
    box_width : float
        Width of each box in x-data units. Default 0.55 (boxes sit
        comfortably within the 1.0-wide category column with room for
        the points underlayer to extend slightly past the box edges).
    order, label_size, xlabel, ylabel, title, log_y, pseudocount,
    origin_zero, ylim, point_size, filename, figsize, ax
        Same semantics as ``pavlab_stripchart``.

    Returns
    -------
    matplotlib Axes
    """
    # ---- validate / coerce inputs --------------------------------------
    x_raw = np.asarray(x, dtype=object).ravel()
    y_arr = np.asarray(y, dtype=float).ravel()
    if len(x_raw) != len(y_arr):
        raise ValueError(
            f"x and y must have equal length ({len(x_raw)} vs {len(y_arr)})"
        )

    color_raw = np.asarray(color) if color is not None else None
    if color_raw is not None and color_raw.ndim == 1 and len(color_raw) != len(x_raw):
        raise ValueError(
            f"color array length ({len(color_raw)}) must match x/y ({len(x_raw)})"
        )

    pk = points_kind.lower() if isinstance(points_kind, str) else "swarm"
    if pk not in ("strip", "swarm"):
        raise ValueError(f"points_kind must be 'strip' or 'swarm', got {points_kind!r}")

    # ---- drop non-finite y ---------------------------------------------
    finite_mask = np.isfinite(y_arr)
    if not finite_mask.all():
        x_raw = x_raw[finite_mask]
        y_arr = y_arr[finite_mask]
        if color_raw is not None and color_raw.ndim == 1 and len(color_raw) == len(finite_mask):
            color_raw = color_raw[finite_mask]

    if len(y_arr) == 0:
        warnings.warn("pavlab_boxplot: no finite data points to plot.", stacklevel=2)

    # ---- resolve group order -------------------------------------------
    seen_groups: list[str] = list(dict.fromkeys(x_raw.tolist()))
    if order is not None:
        unique_groups = []
        for g in order:
            if g not in seen_groups:
                warnings.warn(
                    f"pavlab_boxplot: group {g!r} listed in order has no "
                    "data; skipping.",
                    stacklevel=2,
                )
            else:
                unique_groups.append(g)
        group_set = set(unique_groups)
        keep = np.array([g in group_set for g in x_raw.tolist()], dtype=bool)
        x_raw = x_raw[keep]
        y_arr = y_arr[keep]
        if color_raw is not None and color_raw.ndim == 1:
            color_raw = color_raw[keep]
    else:
        unique_groups = seen_groups

    n = len(y_arr)

    # ---- log transform (matches pavlab_stripchart) ---------------------
    if log_y is False:
        _maybe_warn_log(y_arr, "y")
    y_kind = _pick_log_kind(log_y, y_arr, "y")
    if y_kind:
        neg = int(np.sum(np.isfinite(y_arr) & (y_arr + pseudocount <= 0)))
        if neg:
            warnings.warn(
                f"pavlab_boxplot: {neg} y value(s) will produce -inf after "
                f"log transform (pseudocount={pseudocount} insufficient).",
                stacklevel=2,
            )
        y_arr = _apply_log(y_arr, y_kind, pseudocount)

    # ---- figure setup --------------------------------------------------
    own_fig = ax is None
    if own_fig:
        fig, ax = plt.subplots(figsize=figsize or (5.0, 5.0))
    else:
        fig = ax.figure

    ax.set_facecolor("white")
    ax.grid(False)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)

    label_pt, tick_pt = _font_sizes(label_size)

    # Lock x/y limits BEFORE drawing so the swarm transform is meaningful.
    # We can't rely on matplotlib's auto-fit (nothing's been drawn yet),
    # so compute ylim from the data directly with a small margin.
    group_index = {g: i for i, g in enumerate(unique_groups)}
    ax.set_xlim(-0.5, len(unique_groups) - 0.5)
    if ylim is not None:
        ax.set_ylim(ylim)
    else:
        fy = y_arr[np.isfinite(y_arr)]
        if fy.size > 0:
            y_lo = float(fy.min())
            y_hi = float(fy.max())
            margin = max(1e-9, 0.05 * (y_hi - y_lo)) if y_hi > y_lo else 1.0
            if origin_zero and y_lo >= 0:
                ax.set_ylim(0, y_hi + margin)
            else:
                ax.set_ylim(y_lo - margin, y_hi + margin)

    # ---- per-group data --------------------------------------------------
    groups_arr = np.array([group_index[g] for g in x_raw.tolist()], dtype=int)
    per_group_y = [y_arr[groups_arr == gi] for gi in range(len(unique_groups))]

    # ---- raw points underneath (lowest zorder) -------------------------
    if show_points and n > 0:
        ps = float(point_size) if point_size is not None else (
            20.0 if n <= _N_ALPHA else 10.0
        )
        # Build per-point x position.
        x_pts = np.array(groups_arr, dtype=float)
        if pk == "swarm":
            for gi in range(len(unique_groups)):
                mask = groups_arr == gi
                if not np.any(mask):
                    continue
                offs = _swarm_x_offsets(y_arr[mask], ax, ps)
                # Clip swarm offsets to box_width so points stay roughly
                # within the box's horizontal footprint.
                clip = box_width * 0.45
                offs = np.clip(offs, -clip, clip)
                x_pts[mask] = gi + offs
        else:
            if jitter > 0:
                x_pts = x_pts + np.random.uniform(-jitter, jitter, size=n)

        # Resolve point colour. Mirrors pavlab_stripchart's colour logic.
        if color is None:
            c_arg, cmap_arg, norm_arg, legend_handles, is_single = (
                _POINT_COLOR, None, None, None, True,
            )
        elif isinstance(color, str):
            c_arg, cmap_arg, norm_arg, legend_handles, is_single = (
                color, None, None, None, True,
            )
        else:
            color_arg = (
                color_raw if (color_raw is not None and color_raw.ndim == 1)
                else color
            )
            c_arg, cmap_arg, norm_arg, legend_handles, is_single = _resolve_color(
                color_arg, n,
            )

        if n <= _N_ALPHA:
            alpha = 1.0
        else:
            frac = min(1.0, (n - _N_ALPHA) / max(1, 10_000 - _N_ALPHA))
            alpha = max(0.15, 0.6 - frac * 0.45)

        sc_kw: dict[str, Any] = {
            "s": ps,
            "alpha": alpha,
            "linewidths": 0,
            "marker": "o",
            "zorder": 1,  # below the box
        }
        if is_single:
            sc_kw["color"] = c_arg
        else:
            sc_kw["c"] = c_arg
            if cmap_arg is not None:
                sc_kw["cmap"] = cmap_arg
            if norm_arg is not None:
                sc_kw["norm"] = norm_arg
        sc = ax.scatter(x_pts, y_arr, **sc_kw)
        if cmap_arg is not None:
            cb = fig.colorbar(sc, ax=ax)
            if color_label:
                cb.set_label(color_label, fontsize=tick_pt)
        if legend_handles:
            ax.legend(handles=legend_handles, frameon=False, fontsize=tick_pt)

    # ---- the box itself ------------------------------------------------
    # Skip empty groups (matplotlib boxplot dislikes empty arrays).
    positions = []
    box_data = []
    for gi, ys in enumerate(per_group_y):
        if ys.size > 0:
            positions.append(gi)
            box_data.append(ys)

    if box_data:
        bp = ax.boxplot(
            box_data,
            positions=positions,
            widths=box_width,
            showfliers=False,           # raw points already show outliers
            patch_artist=True,
            medianprops=dict(color=_BOX_COLOR, linewidth=_MEDIAN_LW),
            boxprops=dict(facecolor=_BOX_FILL, edgecolor=_BOX_COLOR,
                          linewidth=_BOX_LW, alpha=0.85),
            whiskerprops=dict(color=_BOX_COLOR, linewidth=_WHISKER_LW),
            capprops=dict(color=_BOX_COLOR, linewidth=_CAP_LW),
            zorder=2,
        )
        # boxplot's per-artist zorder kwarg is honored for the patch but
        # not for whiskers/medians in older mpl. Belt-and-braces:
        for line in bp["whiskers"] + bp["medians"] + bp["caps"]:
            line.set_zorder(2.5)
        for patch in bp["boxes"]:
            patch.set_zorder(2)

    # ---- re-apply x-tick labels (matplotlib's boxplot overrides them) --
    ax.set_xticks(range(len(unique_groups)))
    ax.set_xticklabels(unique_groups)

    # ---- axis labels + title -------------------------------------------
    xl = xlabel or ""
    yl = _axis_label(ylabel, y_kind)
    if xl:
        ax.set_xlabel(xl, fontsize=label_pt)
    if yl:
        ax.set_ylabel(yl, fontsize=label_pt)
    if title:
        ax.set_title(title, fontsize=label_pt, loc="left", fontweight="normal")
    ax.tick_params(axis="both", labelsize=tick_pt)

    # ---- save ----------------------------------------------------------
    if own_fig:
        fig.tight_layout()
        if filename is not None:
            fig.savefig(filename, dpi=200)
            plt.close(fig)

    return ax

File: skills/plotting/R/pavlab_boxplot.R
# pavlab_boxplot: opinionated lab-style boxplot (ggplot2 backend).
#
# Used sparingly. The box itself — quartiles + whiskers + median — IS
# the spread/uncertainty visualization; no separate overlaid error bar
# is drawn. Two modes:
#
#   show_points=TRUE  (default): raw data points (swarmed) under the box.
#   show_points=FALSE:           compact box only (multi-panel layouts
#                                where visual compactness matters).
#
# There is intentionally no `error=` parameter. A separate mean +/- SD
# or +/- 95% CI overlay would double the spread cue the box already
# carries via its whiskers and IQR. Helpers for parameter-estimate
# plots (forest plots, regression-coefficient charts) WILL carry an
# error-bar API in the future, following the cross-plot rule in
# feedback_error_bars_never_sem.md — SD for raw-data charts, 95% CI
# for parameter-estimate charts, never SEM.
#
# Helpers reused from pavlab_scatter.R (which sources palettes.R):
#   .pick_log_kind  .apply_log  .axis_label  .font_sizes  .maybe_warn_log
#   .color_mode  .N_ALPHA  .ACCENT_COLORS  .LABEL_SIZES
#
# Usage:
#   source("R/palettes.R")
#   source("R/pavlab_scatter.R")
#   source("R/pavlab_boxplot.R")
#   pavlab_boxplot(groups, values)
#   pavlab_boxplot(groups, values, error = "sd")
#   pavlab_boxplot(groups, values, points_kind = "strip")
#   pavlab_boxplot(groups, values, show_points = FALSE, error = NULL)  # warns

# ---- Main function ---------------------------------------------------------

#' Pavlab-style boxplot with raw-points underlayer.
#'
#' Categorical x-axis, continuous y-axis. Default rendering: pale box
#' with structural-coloured outline + raw points (swarmed by default)
#' underneath. No separate error-bar overlay — the box's whiskers and
#' IQR are the spread visualisation. Set `show_points=FALSE` for a
#' compact box-only chart.
#'
#' @param x Character or factor vector of group labels (length n).
#' @param y Numeric vector of values (length n). Non-finite values are
#'   silently dropped.
#' @param color NULL, single color string, character/factor vector, or
#'   numeric vector — same semantics as pavlab_stripchart. Affects the
#'   points layer only; the box always uses the structural lab colour.
#' @param color_label Colorbar title for numeric `color`.
#' @param show_points TRUE (default) draws raw points underneath the box.
#' @param points_kind "swarm" (default) or "strip" — layout for the points
#'   underlayer. "swarm" requires `ggbeeswarm`.
#' @param jitter Half-width of jitter when `points_kind="strip"`. Default 0.2.
#' @param box_width Width of each box in x-data units. Default 0.55.
#' @param order Character vector of group levels controlling x-axis order.
#' @param log_y FALSE | TRUE | "log2" | "log10". Auto-selects when TRUE.
#' @param pseudocount Added to y before log transform. Default 1.
#' @param origin_zero TRUE pins non-negative y lower bound to 0.
#' @param ylim Explicit y limits (length 2).
#' @param label_size "small" / "med" / "large" / numeric pt size.
#' @param xlabel,ylabel,title Plot labels.
#' @param point_size Override marker size in mm.
#' @param filename Save path; format inferred from extension.
#' @param figsize Numeric length-2 c(width, height) in inches.
#' @return ggplot2 object (invisibly).
#' @export
pavlab_boxplot <- function(
  x, y,
  color       = NULL,
  color_label = NULL,
  show_points = TRUE,
  points_kind = "swarm",
  jitter      = 0.2,
  box_width   = 0.55,
  order       = NULL,
  log_y       = FALSE,
  pseudocount = 1.0,
  origin_zero = TRUE,
  ylim        = NULL,
  label_size  = "med",
  xlabel      = NULL,
  ylabel      = NULL,
  title       = NULL,
  point_size  = NULL,
  filename    = NULL,
  figsize     = NULL
) {
  if (!requireNamespace("ggplot2", quietly = TRUE))
    stop("pavlab_boxplot: ggplot2 is required. Install with install.packages('ggplot2').",
         call. = FALSE)

  points_kind <- match.arg(tolower(as.character(points_kind)),
                           c("strip", "swarm"))
  if (show_points && points_kind == "swarm" &&
      !requireNamespace("ggbeeswarm", quietly = TRUE)) {
    stop("pavlab_boxplot: points_kind='swarm' requires the ggbeeswarm package. ",
         "Install with install.packages('ggbeeswarm'), or pass points_kind='strip'.",
         call. = FALSE)
  }

  x <- as.character(x)
  y <- as.numeric(y)
  if (length(x) != length(y))
    stop(sprintf("x and y must have equal length (%d vs %d)", length(x), length(y)),
         call. = FALSE)

  has_vec_color <- !is.null(color) && length(color) == length(y)
  mask <- is.finite(y)
  if (has_vec_color) color <- color[mask]
  x <- x[mask]; y <- y[mask]
  n <- length(y)
  if (n == 0) warning("pavlab_boxplot: no finite data points to plot.", call. = FALSE)

  # Suggest log scale when not transforming.
  if (identical(log_y, FALSE)) .maybe_warn_log(y, "y")
  y_kind <- .pick_log_kind(log_y, y, "y")
  if (!is.null(y_kind)) {
    neg <- sum(is.finite(y) & (y + pseudocount) <= 0)
    if (neg > 0)
      warning(sprintf(
        "pavlab_boxplot: %d y value(s) will produce -Inf after log transform (pseudocount=%g insufficient).",
        neg, pseudocount), call. = FALSE)
    y <- .apply_log(y, y_kind, pseudocount)
  }

  seen <- unique(x)
  if (!is.null(order)) {
    extra  <- seen[!seen %in% order]
    levels <- c(order, extra)
  } else {
    levels <- seen
  }
  x_fac <- factor(x, levels = levels)

  fs       <- .font_sizes(label_size)
  label_pt <- fs[1]
  tick_pt  <- fs[2]
  yl       <- .axis_label(ylabel, y_kind)

  ps <- if (!is.null(point_size)) as.numeric(point_size) else
          if (n <= .N_ALPHA) 1.7 else 1.3
  alpha_val <- if (n <= .N_ALPHA) 0.85 else {
    frac <- min(1.0, (n - .N_ALPHA) / max(1L, 10000L - .N_ALPHA))
    max(0.15, 0.6 - frac * 0.45)
  }

  df <- data.frame(x_grp = x_fac, y_val = y, stringsAsFactors = FALSE)

  # ---- Style constants (match the Python helper) --------------------------
  BOX_COLOR   <- "#1f2937"
  POINT_COLOR <- "#374151"
  BOX_FILL    <- "#f3f4f6"

  # Base aesthetic
  cmode <- .color_mode(color, n)
  if (cmode %in% c("uniform_black", "uniform_string")) {
    p <- ggplot2::ggplot(df, ggplot2::aes(x = x_grp, y = y_val))
  } else if (cmode == "categorical") {
    df$color_col <- as.character(color)
    p <- ggplot2::ggplot(df, ggplot2::aes(x = x_grp, y = y_val))
  } else {
    df$color_col <- as.numeric(color)
    p <- ggplot2::ggplot(df, ggplot2::aes(x = x_grp, y = y_val))
  }

  # ---- Layer 1: raw points underneath ------------------------------------
  if (show_points) {
    point_layer <- function(...) {
      if (points_kind == "swarm") {
        ggbeeswarm::geom_beeswarm(..., cex = 1.0, method = "swarm",
                                  shape = 16, stroke = 0, size = ps)
      } else {
        ggplot2::geom_jitter(..., shape = 16, stroke = 0, size = ps,
                             position = ggplot2::position_jitter(width = jitter, height = 0))
      }
    }
    if (cmode == "uniform_black") {
      p <- p + point_layer(color = POINT_COLOR, alpha = alpha_val)
    } else if (cmode == "uniform_string") {
      p <- p + point_layer(color = as.character(color), alpha = alpha_val)
    } else if (cmode == "categorical") {
      cats     <- unique(as.character(color))
      cat_idx  <- ((seq_along(cats) - 1L) %% length(.ACCENT_COLORS)) + 1L
      cat_cols <- setNames(.ACCENT_COLORS[cat_idx], cats)
      p <- p +
        point_layer(ggplot2::aes(color = color_col), alpha = alpha_val) +
        ggplot2::scale_color_manual(values = cat_cols, name = NULL) +
        ggplot2::guides(color = ggplot2::guide_legend(
          override.aes = list(size = 3, alpha = 1)))
    } else {
      clabel <- if (!is.null(color_label)) color_label else ""
      p <- p +
        point_layer(ggplot2::aes(color = color_col), alpha = alpha_val) +
        ggplot2::scale_color_viridis_c(name = clabel)
    }
  }

  # ---- Layer 2: the box itself --------------------------------------------
  p <- p + ggplot2::geom_boxplot(
    fill      = BOX_FILL,
    color     = BOX_COLOR,
    alpha     = 0.85,
    width     = box_width,
    outlier.shape = NA,       # raw points show outliers already
    linewidth = 0.55,
    fatten    = 1.3            # median line slightly heavier than box outline
  )

  # ---- Theme + labels + limits -------------------------------------------
  p <- p +
    ggplot2::theme_classic(base_size = tick_pt) +
    ggplot2::theme(
      axis.title        = ggplot2::element_text(size = label_pt, color = "#1f2937"),
      axis.text         = ggplot2::element_text(size = tick_pt,  color = "#6b7280"),
      axis.ticks        = ggplot2::element_line(color = "#6b7280"),
      plot.title        = ggplot2::element_text(size = label_pt, color = "#1f2937",
                                                hjust = 0, face = "plain"),
      legend.text       = ggplot2::element_text(size = tick_pt),
      legend.title      = ggplot2::element_text(size = tick_pt),
      legend.background = ggplot2::element_blank(),
      legend.key        = ggplot2::element_blank(),
      plot.margin       = ggplot2::margin(3, 3, 3, 3, "mm")
    )

  if (!is.null(xlabel) && nchar(xlabel) > 0) p <- p + ggplot2::xlab(xlabel)
  if (nchar(yl) > 0) p <- p + ggplot2::ylab(yl)
  if (!is.null(title)) p <- p + ggplot2::ggtitle(title)

  if (origin_zero) {
    fy <- df$y_val[is.finite(df$y_val)]
    if (length(fy) > 0 && min(fy) >= 0) p <- p + ggplot2::expand_limits(y = 0)
  }
  if (!is.null(ylim)) p <- p + ggplot2::coord_cartesian(ylim = ylim)

  if (!is.null(filename)) {
    w <- if (!is.null(figsize)) figsize[1] else 5.5
    h <- if (!is.null(figsize)) figsize[2] else 5.0
    ggplot2::ggsave(filename, plot = p, width = w, height = h, dpi = 150)
  }

  invisible(p)
}
